### ISSUE TITLE
How to make A2C learn for reward that does not go negative at all?

### ISSUE THREAD

**Issue by Harishu1998** [DLR-RM/stable-baselines3#1583]
### ❓ Question

I have a custom environment which is a simulation of a fermentation process. The fermentation can run for 1200 steps only (length of experiment). There are 2 termination steps for my environment which is

at the end of experiment ie. experiment_time_steps >=1200
as soon as the product concentration starts decreasing.
when I run my custom environment on A2C using model.learn(total_timesteps = 10,000) I observe the environment reset() while total_timesteps counter keeps increasing. Ex.

```
total_timesteps = 1 , experiement_time_steps = 1 , reward = 10 
total_timesteps = 2 , experiement_time_steps = 2 , reward = 10
total_timesteps = 3 , experiement_time_steps = 3 , reward = 10
total_timesteps = 4 , experiement_time_steps = 4 , reward = 10
**** product declined 
total reward 40 (I assume they add each steps() reward that is returned)
reset()
total_timesteps = 5 , experiement_time_steps = 1 , reward = 10
total_timesteps = 6 , experiement_time_steps = 2 , reward = 10
total_timesteps = 7 , experiement_time_steps = 3 , reward = 10
total_timesteps = 8 , experiement_time_steps = 4 , reward = 10
total_timesteps = 9 , experiement_time_steps = 5 , reward = 10
total_timesteps = 10 , experiement_time_steps = 6 , reward = 10
total_timesteps = 11 , experiement_time_steps = 7 , reward = 10
product declined
total_reward = 70
reset()
```
Because my termination due to enzyme decline is completely random, it sometimes runs for 200 times (the step function) and the reward value is around 500 but sometimes it runs only for 100 iterations because the actions taken caused the product to decline early and my reward value is 50.

So My plot for ep_mean(tensorboard) is random lines going up and down which is not something I am expecting because I want to see the model learning and the ep mean going up. Is there a way I can discard the low ep_mean runs and plot only for high rewards/increasing rewards?

Ex. out of 10,000 timesteps my environment resets 5 times due to decline in product. so I get 5 different reward values at the end like below

```
total_timesteps = 1 - 500 , reward = 700
total_timesteps = 500-700 , reward = 200
total_timesteps = 700-4000, reward = 1000
total_timesteps = 4001-9000 reward = 100
total_timesteps = 9001-10000, reward = 10
```
when this is plotted I get a graph that is fluctuating.

So, Basically my A2C model is not learning right? Because before each reset the reward function is giving out different values(high and low but +ve).

How can I make my model learn in such a situation?

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by qgallouedec** [comment]
> while total_timesteps counter keeps increasing

`total_timesteps` is the number of training steps, so it's expected.

>  I want to see the model learning and the ep mean going up

If the length of the episode doesn't depend on the agent's actions, then you need to consider the average total reward over several episodes. If the agent learns properly, then this value is expected to increase.

> Is there a way I can discard the low ep_mean runs and plot only for high rewards/increasing rewards?

No, for the simple reason that it makes no sense to consider only episodes where the total reward is high, since the aim of the RL is to optimize the expected total reward. 
However, if you think it really makes sense for you, I suggest you modify the sb3 source code to support this feature.

> So, Basically my A2C model is not learning right? Because before each reset the reward function is giving out different values(high and low but +ve).

Once again, the RL optimizes the expected total reward. The stochasticity of the environment can lead to variability in the total reward, but what's important is that this value increases on average. So no, you can't conclude that the agent doesn't learn just because the total reward varies.
